data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,392 @@ category: "Module System"
+```ocaml
+module StringSet = Set.Make(String)
+```
+
+With this, the command `dune exec funkt` shouldn't do anything, but it shouldn't fail either. Here is the meaning of that statement:
+- The module `funkt.StringSet` is defined.
+- The module `String` (from the standard library) is applied to the functor `Set.Make`. This is allowed because the `String` module satisfies the interface `Set.OrderedSet`.
+  - It defines a type name `t` (which is an alias for `string`).
+  - It defines a function `compare` of type `t -> t -> bool`, that is the function `String.compare`.
+- The result module from the functor application `Set.Make(String)` is bound to the name `StringSet`, and it has the signature `Set.S`.
+
+Add some code to the `funkt.ml` file to produce an executable that does something and checks the result.
+
+**`funkt.ml`**
+```ocaml
+module StringSet = Set.Make(String)
+
+let _ =
+  In_channel.input_lines stdin
+  |> List.concat_map Str.(split (regexp "[ \t.,;:()]+"))
+  |> StringSet.of_list
+  |> StringSet.iter print_endline
+```
+
+Here are the types of functions used throughout the pipe:
+- `In_channel.input_lines : in_channel -> string list`,
+- `Str.(split (regexp "[ \t.,;:()]+")) : string -> string list`,
+- `List.concat_map : ('a -> 'b list) -> 'a list -> 'b list`,
+- `StringSet.of_list : string list -> StringSet.t`, and
+- `StringSet.iter : StringSet.t -> unit`.
+
+This reads the following way:
+- Read lines of text from standard input, produce a list of strings.
+- Split each string using a regular expression, flatten the list of lists into a list.
+- Convert the list of strings into a set.
+- Display each element of the set.
+
+The functions `StringSet.of_list` and `StringSet.iter` are available as the result of the functor application.
+
+```shell
+$ dune exec funkt < dune
+executable
+libraries
+name
+public_name
+str
+funkt
+```
+
+There are no duplicates in a `Set`. Therefore, the string `"funkt"` is only displayed once, although it appears twice in the `dune` file.
+
+## Extending a Module with a Standard Library Functor
+
+Using the `include` statement, here is an alternate way to expose the module created by `Set.Make(String)`:
+
+**`funkt.ml`**
+```ocaml
+module String = struct
+  include String
+  module Set = Set.Make(String)
+end
+
+let _ =
+  stdin
+  |> In_channel.input_lines
+  |> List.concat_map Str.(split (regexp "[ \t.,;:()]+"))
+  |> String.Set.of_list
+  |> String.Set.iter print_endline
+```
+
+This allows the user to seemingly extend the module `String` with a submodule `Set`. Check the behaviour using `dune exec funkt < dune`.
+
+## Functors are Parametrised Modules
+
+### Functors from the Standard Library
+
+Some ”modules” provide operations over an abstract type and need to be supplied with a parameter module used in their implementation. These “modules” are parametrised, in other words, functors. That's the case for the sets, maps, and hash tables provided by the standard library. It works in a contract way:
```

Comment:
What do you mean here by "contract way"? I'm unfamiliar with that term. 
Might you mean "a specific" or "specified" way? Like, it works this way every time?

Reply:
Is the new text an improvement?

Reply:
Yes! Much more clear!